internals/overlord/servstate/workloads.go

```diff
@@ -0,0 +1,223 @@
+			}
+		}
+	}
+	for name, workload := range ws.Entries {
+		if _, _, err := osutil.NormalizeUidGid(workload.UserID, workload.GroupID, workload.User, workload.Group); err != nil {
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`%v for workload %q`, err, name),
+			}
+		}
+	}
+	return nil
+}
+
+const WorkloadsField = "workloads"
+
+var _ plan.Section = (*WorkloadsSection)(nil)
+
+type WorkloadsSection struct {
+	Entries map[string]*Workload `yaml:",inline"`
+}
+
+func (ws *WorkloadsSection) IsZero() bool {
+	return len(ws.Entries) == 0
+}
+
+func (ws *WorkloadsSection) Validate() error {
+	for name, workload := range ws.Entries {
+		if workload == nil {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("workload %q has a null value", name),
+			}
+		}
+		if err := workload.validate(); err != nil {
+			return &plan.FormatError{
+				Message: fmt.Sprintf("workload %q %v", name, err),
+			}
+		}
+	}
+	return nil
+}
+
+func (ws *WorkloadsSection) combine(other *WorkloadsSection) error {
+	ws.Entries = makeMapIfNil(ws.Entries)
+	for name, workload := range other.Entries {
+		switch workload.Override {
+		case plan.MergeOverride:
+			if current, ok := ws.Entries[name]; ok {
+				copied := current.copy()
+				copied.merge(workload)
+				ws.Entries[name] = copied
+				break
+			}
+			fallthrough
+		case plan.ReplaceOverride:
+			ws.Entries[name] = workload.copy()
+		case plan.UnknownOverride:
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`workload %q must define an "override" policy`, name),
+			}
+		default:
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`workload %q has an invalid "override" policy: %q`, name, workload.Override),
+			}
+		}
+	}
+	return nil
+}
+
+type Workload struct {
+	// Basic details
+	Name     string        `yaml:"-"`
+	Override plan.Override `yaml:"override,omitempty"`
+
+	// Options for command execution
+	Environment map[string]string `yaml:"environment,omitempty"`
+	UserID      *int              `yaml:"user-id,omitempty"`
+	User        string            `yaml:"user,omitempty"`
+	GroupID     *int              `yaml:"group-id,omitempty"`
+	Group       string            `yaml:"group,omitempty"`
+}
+
+func (w *Workload) validate() error {
+	if w.Name == "" {
+		return errors.New("cannot have an empty name")
+	}
+	// Value of Override is checked in the (*WorkloadSection).combine() method
+	return nil
+}
+
+func (w *Workload) copy() *Workload {
+	copied := *w
+	if w.Environment != nil {
+		copied.Environment = make(map[string]string, len(w.Environment))
+		for k, v := range w.Environment {
+			copied.Environment[k] = v
+		}
+	}
+	if w.UserID != nil {
+		copied.UserID = copyPtr(w.UserID)
+	}
+	if w.GroupID != nil {
+		copied.GroupID = copyPtr(w.GroupID)
+	}
+	return &copied
+}
+
+func (w *Workload) merge(other *Workload) {
+	w.Environment = makeMapIfNil(w.Environment)
+	for k, v := range other.Environment {
+		w.Environment[k] = v
```

Comment:
I'd prefer to do the `makeMapIfNil` inside the loop here (like `Service.Merge` and others do) so that if both maps are nil, the result is nil. Only allocate if we have to.

Or maybe we can use `maps.Copy`:

```go
if len(other.Environment) > 0 {
    w.Environment = makeMapIfNil(w.Environment)
    maps.Copy(w.Environment, other.Environment)
}
```

Then again, that dance is kind of annoying, so maybe a `copyMap` helper that does this. I'll give it some more thought separately to this PR.